### ISSUE TITLE
Suggestion: `elementAtOrNull` should return `null` if the index is negative

### ISSUE THREAD

**Issue by nate-thegrate** [dart-lang/sdk#56185]
```diff
T? elementAtOrNull(int index) {
- RangeError.checkNotNegative(index, "index");
+ if (index < 0) return null;
  if (this is EfficientLengthIterable) {
    if (index >= length) return null;
    return elementAt(index);
  }
  var iterator = this.iterator;
  do {
    if (!iterator.moveNext()) return null;
  } while (--index >= 0);
  return iterator.current;
}
```

This would enable some really nice-looking code:

```dart
for (final (i, child) in widget.children.indexed) {
  if (children.elementAtOrNull(i + 1)?.key case final nextKey?) {
    // works!
  }

  if (children.elementAtOrNull(i - 1)?.key case final prevKey?) {
    // it'd be great if this worked too!
  }
}
```

**Comment by lrhn** [comment]
It will also enable mistakes to be silently ignored.

The reason `elementAtOrNull` doesn't throw for positive numbers is that while there is no element at that position, it's not something the caller can be expected to know without first checking the length of the iterable, which is too expensive to expect them to do.

The reason it throws for negative numbers is that there is never an element with that number, and the caller it's expected to know that.

The `.elementAtOrNull(n)` is basically equivalent to `.skip(n).firstOrNull`, which also throws for negative `n`s.


And rather than this code, just do
`for (var i = 0; i < widget.children.length; i++) {
  var child = widget.children[i];
}

**Comment by nate-thegrate** [comment]
> The reason `elementAtOrNull` doesn't throw for positive numbers is that while there is no element at that position, it's not something the caller can be expected to know without first checking the length of the iterable, which is too expensive to expect them to do.

Yeah, that makes a lot of sense—it'd be a waste of resources to iterate through the whole thing if you're just looking for index `2` or something.

<br>

I almost always work with EfficientLengthIterables, so every time I use `.elementAtOrNull()`, it's for readability rather than efficiency.

<details> <summary>Using <code>[]</code> instead of <code>elementAtOrNull</code> makes the original example a bit more verbose. (click to expand)</summary>

```dart
for (var i = 0; i < widget.children.length; i++) {
  var child = widget.children[i];

  var nextKey = (i + 1 >= widget.children.length) ? null : widget.children[i + 1].key;
  if (nextKey != null) {
    // ...
  }

  var prevKey = (i - 1 < 0) ? null : widget.children[i - 1].key;
  if (prevKey != null) {
    // ...
  }
}
```

</details>

<br>

There's a real-world example of this in [analyze_snippet_code.dart](https://github.com/flutter/flutter/blob/5ebc993dffe8ce7e91479b8ff7b003c1f464a621/dev/bots/analyze_snippet_code.dart#L995-L1008).

```dart
// using []
if ((lineNumber - delta > 0) && (lineNumber - delta <= snippet.code.length) && !snippet.code[lineNumber - delta - 1].generated) {
  actualSource = snippet.code[lineNumber - delta - 1];
  actualLine = actualSource.line;
  break;
}
if ((lineNumber + delta < snippet.code.length) && (lineNumber + delta >= 0) && !snippet.code[lineNumber + delta].generated) {
  actualSource = snippet.code[lineNumber + delta];
  actualLine = actualSource.line;
  break;
}



// using elementAtOrNull
if (snippet.code.elementAtOrNull(lineNumber - delta - 1) case var source? when !source.generated) {
  actualSource = source;
  actualLine = source.line;
  break;
}
if (snippet.code.elementAtOrNull(lineNumber + delta) case var source? when !source.generated) {
  actualSource = source;
  actualLine = source.line;
  break;
}
```

In my opinion, accepting a negative index without throwing an error opens up valuable use cases for the method with very little downside. Now that we have [control flow collections](https://dart.dev/language/collections#control-flow-operators), I've noticed the amount of non-EfficientLengthIterables go down significantly, but perhaps they're still common in other places.

**Comment by lrhn** [comment]
That's an impressive number of run-on conditions, so I can see why it's useful to have a single function doing the range testing and element lookup, but you can do that with a simple helper function too:
```dart
T? tryElement<T extends Object>(List<T> elements, int index) {
  if (index >= 0 && index < elements.length) return elements[index];
  return null;
}

if (tryElement(snippet.code, lineNumber - delta - 1) case var source? when !source.generated) {
  ...
}
...
```
That works for you when you know `elements.length` is efficient, and probably `elementAt` too. (If `elementAt` is not efficient, it's not that great to do `elementAt` in *each* condition, only to maybe throw the result away again if it's `.generated`. It may be more efficient to iterate once from zero and see which of `lineNumber - delta - 1` or `lineNumber + delta` you hit, if any. And if you know it's a list, just use it as a list.)

I know it's annoying to have a method that *almost* solves your use-case, but it wasn't designed for your use-case, so that it works so well is mostly accidental. It's not a good enough reason to change it.
